Add support for NCCL alltoall

zasdfgbnm opened the pull request:
In #42514, NCCL `alltoall_single` is already added. This PR adds NCCL `alltoall`.

The difference between `alltoall_single` and `alltoall` is: `alltoall_single`  works on a single tensor and send/receive slices of that tensor, while `alltoall` works on a list of tensor, and send/receive tensors in that list.

cc: @ptrblck @ngimel

zasdfgbnm committed:
Add support for NCCL all to all

zasdfgbnm committed:
fix

zasdfgbnm committed:
fix

zasdfgbnm committed:
fix

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm committed:
tests

zasdfgbnm committed:
fix

zasdfgbnm committed:
fix

zasdfgbnm committed:
cleanup

ngimel commented:
We already enabled alltoall in #42514, how is this different?

zasdfgbnm commented:
@ngimel That PR only enables `alltoall_single`, not `alltoall`.

ngimel commented:
And what is the difference? It would be good to have at least a brief description of this PR.

zasdfgbnm commented:
@ngimel I updated https://github.com/pytorch/pytorch/pull/44374#issue-482622480

srinivas212 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1164,12 +1164,60 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
+        at::Tensor& /* unused */,
+        ncclComm_t comm,
+        at::cuda::CUDAStream& stream) {
+      for (size_t r = 0; r < outputTensors.size(); r++) {
```
Why not reuse NcclAlltoallv to do the send/recv? 
https://github.com/pytorch/pytorch/blob/37093f4d99c22cb9d89d9eb1cd56b97be92938b4/torch/lib/c10d/ProcessGroupNCCL.cpp#L187

zasdfgbnm replied:
`ncclAlltoallv` requires the arguments to be in a format of base pointers, displacements, and counts. This is good for `all_to_all_single` because we are sending/receiving slices of tensors. But for `all_to_all`, since each tensor in the list independent, converting the arguments into this format is not easier than implementing it with ncclSend/ncclRecv directly.

srinivas212 replied:
@zasdfgbnm thanks for you patience and sorry for the delay. 

One minor change: put the NCCL Send/Recv with a group start/end similar to here: https://github.com/pytorch/pytorch/blob/92921c82bb7c0d4f1874ab354d2aa903b662c355/torch/lib/c10d/ProcessGroupNCCL.cpp#L183

This should not have effect on performance since these operations get encapsulated in a group start/end in the collective: 
https://github.com/pytorch/pytorch/blob/92921c82bb7c0d4f1874ab354d2aa903b662c355/torch/lib/c10d/ProcessGroupNCCL.cpp#L1065

The reason for requesting this change is to give the clear impression that all these sends/recvs are notionally part of a single call.

zasdfgbnm replied:
@srinivas212 Done!

srinivas212 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1164,12 +1164,60 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
+    std::vector<at::Tensor>& inputTensors,
+    const AllToAllOptions& /* unused */) {
+  auto device = outputTensors[0].device();
+  for (size_t r = 0; r < outputTensors.size(); r++) {
```
Reuse existing checks? 
https://github.com/pytorch/pytorch/blob/37093f4d99c22cb9d89d9eb1cd56b97be92938b4/torch/lib/c10d/ProcessGroupNCCL.cpp#L687

https://github.com/pytorch/pytorch/blob/37093f4d99c22cb9d89d9eb1cd56b97be92938b4/torch/lib/c10d/ProcessGroupNCCL.cpp#L696

zasdfgbnm replied:
I think `check_gpu_tensors` checks that all tensors are at different devices, which is different from the use case here where tensors in the list are required to be on the same device.

zasdfgbnm commented:
@srinivas212 any update on this?

zasdfgbnm committed:
group

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

srinivas212 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1555,12 +1555,62 @@ void ProcessGroupNCCL::groupEnd() {
   --ncclActiveGroupCounter_;
 }
 
+#ifdef ENABLE_NCCL_P2P_SUPPORT
```
@zasdfgbnm  can you also move the whole code under this #ifdef into the one above w/ alltoall_single, send, recv? 

https://github.com/pytorch/pytorch/blob/1c39d26db63c1c92b90aa269837c749bf94fde57/torch/lib/c10d/ProcessGroupNCCL.cpp#L1478

zasdfgbnm replied:
done

srinivas212 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1555,12 +1555,62 @@ void ProcessGroupNCCL::groupEnd() {
     std::vector<at::Tensor>& /* unused */,
     std::vector<at::Tensor>& /* unused */,
     const AllToAllOptions& /* unused */) {
   throw std::runtime_error("ProcessGroupNCCL does not support alltoall");
```
Once you move this into the ifdef ENABLE_NCCL_P2P_SUPPORT above, you should also change the else error message...  
"ProcessGroupNCCL only supports recv for NCCL lib version >= 2.7.0"

zasdfgbnm replied:
done

zasdfgbnm committed:
save

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm committed:
error message

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm committed:
OpType::ALLTOALL

zasdfgbnm commented:
@srinivas212 I fixed a compilation error at https://github.com/pytorch/pytorch/pull/44374/commits/2ab4ff2e6e394dfa6a4fdb187cb6ec3a513d85c1, and this PR should be ready.

zasdfgbnm commented:
Should I do the same thing as https://github.com/pytorch/pytorch/pull/45900?

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

srinivas212 commented:
> Should I do the same thing as #45900?

@zasdfgbnm please don't use NCCL send/recv directly

https://github.com/pytorch/pytorch/pull/45926/files#diff-43cb0f438d3eb35dec0a1680ddc2d01c3ae9277d91aca4c2119d0b9ea80adeb6R1445

zasdfgbnm committed:
update all to all

zasdfgbnm committed:
more

zasdfgbnm committed:
fix

zasdfgbnm committed:
fix

zasdfgbnm committed:
@skip_if_rocm

zasdfgbnm commented:
@srinivas212 
- I renamed the original `all2all` in `nccl.cpp` into `all2all_single_equal_split`, and
- moved the `ncclAlltoallv` (which uses ncclSend` and `ncclRecv`) from `ProcessGroupNCCL.cpp` to `nccl.cpp` and rename it into `all2all_single_unequal_split`, and
- moved the core of `ProcessGroupNCCL::alltoall` to `nccl.cpp` and name it as `all2all`, and
- enabled all the `all_to_all` tests on the loss CI

Now all the tests are passing, and this PR should be ready

zasdfgbnm committed:
Merge branch 'master' into nccl-all2all

zasdfgbnm committed:
fix

zasdfgbnm commented:
ping @srinivas212

zasdfgbnm committed:
Merge branch 'master' into nccl-all2all

zasdfgbnm commented:
ping @srinivas212

zasdfgbnm committed:
Merge branch 'master' into nccl-all2all

ptrblck commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1503,23 +1460,50 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
   }
 }
 
+std::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall(
```
This is causing the current failure, so `c10::intrusive_ptr<>` should be needed here.

ptrblck commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1568,6 +1552,14 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
       "ProcessGroupNCCL only supports alltoall* for NCCL lib version >= 2.7.0");
 }
 
+std::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall(
```
Same as above: `c10::intrusive_ptr<>`

ptrblck commented:
https://github.com/pytorch/pytorch/blob/e94b60223a0cb52908b973de6d9e0d9e38a07737/torch/lib/c10d/ProcessGroupNCCL.cpp#L1594-L1599

should create a conflict by redefining `ProcessGroupNCCL::alltoall` from the `#ifdef ENABLE_NCCL_P2P_SUPPORT` block in https://github.com/pytorch/pytorch/blob/e94b60223a0cb52908b973de6d9e0d9e38a07737/torch/lib/c10d/ProcessGroupNCCL.cpp#L1481
or
https://github.com/pytorch/pytorch/blob/e94b60223a0cb52908b973de6d9e0d9e38a07737/torch/lib/c10d/ProcessGroupNCCL.cpp#L1555
so I think it should be removed.

zasdfgbnm committed:
Update ProcessGroupNCCL.cpp

zasdfgbnm committed:
Update ProcessGroupNCCL.cpp

srinivas212 commented:
> should create a conflict by redefining `ProcessGroupNCCL::alltoall` from the `#ifdef ENABLE_NCCL_P2P_SUPPORT` block in

This was removed in the following PR: 
https://github.com/pytorch/pytorch/pull/45900

So yes, we need to remove this.

zasdfgbnm commented:
@srinivas212 I already removed it.

mrshenli commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -71,11 +71,8 @@ torch::cuda::nccl::ncclResult from_nccl_result(ncclResult_t var) {
-  switch (t.scalar_type()) {
+ncclDataType_t to_nccl_data_type(c10::ScalarType type) {
+  switch (type) {
     case at::kFloat:
```
Today's `ProcessGroupNCCL` also supports `at::kBool`, is that the same as `at::kByte`?

```
    {at::kChar, ncclInt8},
    {at::kByte, ncclUint8},
    {at::kFloat, ncclFloat},
    {at::kDouble, ncclDouble},
    {at::kInt, ncclInt32},
    {at::kLong, ncclInt64},
    {at::kHalf, ncclHalf},
    {at::kBool, ncclUint8},
#if defined(__HIP_PLATFORM_HCC__) && HIP_VERSION >= 301
    {at::kBFloat16, ncclBfloat16},
#endif
```

zasdfgbnm replied:
I have added `kBool`. And yes, I think kByte should be `ncclUint8` as well, instead of `ncclChar` as currently in this file. I have updated this.

mrshenli commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -99,6 +96,13 @@ ncclDataType_t to_nccl_data_type(const at::Tensor& t) {
 
+ncclDataType_t to_nccl_data_type(const at::Tensor& t) {
+  if (!t.is_cuda()) {
+    throw std::runtime_error("Unconvertible NCCL type");
```
nit: this is prior to this PR, shall we be more explicit on the error message? Should this be the following? 

```
f"NCCL only supports CUDA tensors, but got a tensor on {t.device}"
```

mrshenli commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -99,6 +96,13 @@ ncclDataType_t to_nccl_data_type(const at::Tensor& t) {
   }
```
nit: this is prior to this PR, shall we also tell the type in the error message?

zasdfgbnm replied:
fixed

mrshenli commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -625,7 +629,7 @@ void all_gather(
 }
 
-void all2all(at::Tensor& input,
+void all2all_single_equal_split(at::Tensor& input,
```
Would I be correct if I assume this API is not visible to users?

zasdfgbnm replied:
It seems so? I can not find anything about nccl at https://pytorch.org/cppdocs/api/library_root.html

mrshenli commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -660,6 +664,98 @@ void all2all(at::Tensor& input,
+  NCCL_CHECK(ncclGroupStart());
+  for (int r = 0; r < numranks; r++) {
+    // NCCL uses 0 byte message for synchronization
+    // Avoid send/recv when message size is zero
```
Does this mean even if all send/recv cnts are 0, this would still trigger an zero-byte message to do sync across ranks?

zasdfgbnm replied:
If all send counts are zero, wouldn't this be an empty nccl group?

mrshenli reviewed (approved):
LGTM! Thanks for fixing!!

mrshenli commented:
cc @agolynski

zasdfgbnm committed:
fix

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm committed:
fix

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm commented:
Is there anything this PR is waiting for?

zasdfgbnm committed:
Merge branch 'master' of github.com:pytorch/pytorch into nccl-all2all

zasdfgbnm commented:
Rebased today

zasdfgbnm commented:
ping @srinivas212 any update on this?

cdzhan commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1501,23 +1458,50 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall_base(
+  return collective(
+    inputTensor0,
+    outputTensor0,
+    [&](at::Tensor& /* unused */,
```
@zasdfgbnm Hello, I'm a bit confused, why outputTensors didn't need to record ncclStream to prevent being freed before the collective finishes?

zasdfgbnm replied:
This looks like a bug... Thanks for catching it!

cdzhan replied:
You’re welcome :)
